feat(graph): add Spanner Graph support to LangChain GraphStore interface

mtyin opened the pull request:
1) Add SpannerGraphStore implementation of GraphStore;
2) Add an integration test;
3) Add a notebook to demonstrate the usage;
4) Misc updates: requirements, __init__

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/langchain-google-spanner-python/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

mtyin committed:
feat(graph): add Spanner Graph support to LangChain GraphStore interface

1) Add SpannerGraphStore implementation of GraphStore;
2) Add an integration test;
3) Add a notebook to demonstrate the usage;
4) Misc updates: requirements, __init__

mtyin committed:
Allow more flexible input that can generated by LLMs

1) same edge label between different types of nodes

E.g.

Company FOCUS_ON Product
Company FOCUS_ON Technology

2) the same doc contains multiple nodes/edges with the same key fields.

E.g.

Merge
  Company(id='Google', properties={'p1':'a1', 'p3':'a3'}),
  Company(id='Google', properties={'p2':'a2', 'p3':'a4'})
into
  Company(id='Google', properties={'p1':'a1', 'p2':'a2', 'p3':'a4'})

This is due to Spanner DML doesn't support updating the same row twice
in the same DML. We could also separate each node/edge into a separate
DML which can be very slow.

gauravpurohit06 committed:
Merge branch 'main' into graph

gauravpurohit06 committed:
style: formatting

gauravpurohit06 committed:
style: formatting

gauravpurohit06 commented on src/langchain_google_spanner/graph_store.py:
```diff
@@ -0,0 +1,1176 @@
+            w = NodeWrapper(node)
+            if w in s:
+                # Combine the properties for nodes with the same id.
+                n = next(filter(lambda v: v == w, s))
```
This operation potentially be slow for big graphs, because it's iterating over a complete set which is not required as the `NodeWrapper` object is hashable & comparable based on their ids.

mtyin replied:
Done!

gauravpurohit06 commented on src/langchain_google_spanner/graph_store.py:
```diff
@@ -0,0 +1,1176 @@
+            w = NodeWrapper(node)
+            if w in s:
+                # Combine the properties for nodes with the same id.
+                n = next(filter(lambda v: v == w, s))
```
```
        if w.node.id in s:
            # Combine the properties for nodes with the same id.
            s[w.node.id].node.properties.update(node.properties) 
 ```

mtyin replied:
Done!

mtyin replied:
Btw, I used the wrapper as the key to make the code looks more consistent between node and edge.

gauravpurohit06 commented on src/langchain_google_spanner/graph_store.py:
```diff
@@ -0,0 +1,1176 @@
+            w = EdgeWrapper(edge)
+            if w in s:
+                # Combine the properties for edges with the same id.
+                e = next(filter(lambda v: v == w, s))
```
same as above.

mtyin replied:
Done!

gauravpurohit06 commented on src/langchain_google_spanner/graph_store.py:
```diff
@@ -0,0 +1,1176 @@
+    }
+
+
+class TypeUtility(object):
```
Add class docs.

mtyin replied:
Done! Also moved the class to a separate file

gauravpurohit06 commented on src/langchain_google_spanner/graph_store.py:
```diff
@@ -0,0 +1,1176 @@
+        """
+        if s == "BOOL":
+            return param_types.BOOL
+        if s in ["INT64", "INT32"]:
```
why would the schema have `INT32` ?

mtyin replied:
Good catch!

gauravpurohit06 commented on src/langchain_google_spanner/graph_store.py:
```diff
@@ -0,0 +1,1176 @@
+    }
+
+
+class TypeUtility(object):
```
This is very generic class and not limited just to GraphStore.  Can we create a utility module and move it there? It would be useful for future development in this repo.

mtyin replied:
Done!

gauravpurohit06 commented on src/langchain_google_spanner/graph_store.py:
```diff
@@ -0,0 +1,1176 @@
+from requests.structures import CaseInsensitiveDict
+
+
+def to_identifier(s: str) -> str:
```
Can we move this method in `GraphDocumentUtility`?  as we are keeping `fixup_identifier` method there as well, both methods looks similar to me.

mtyin replied:
Done!

gauravpurohit06 commented on src/langchain_google_spanner/graph_store.py:
```diff
@@ -0,0 +1,1176 @@
+    return "`" + s + "`"
+
+
+def to_identifiers(s: List[str]) -> str:
```
same as above.

mtyin replied:
Done!

gauravpurohit06 reviewed (changes requested):
Thank you [mtyin](https://github.com/mtyin) for raising the PR. The code looks well organized and modular. 

I have reviewed the source code file and not test file. From the functionality point of view, it looks good but I need to do one more iteration. 

Other than mentioned comments, few general things: 
1. You can use `isort` and `black` to format and style the code. 
2. Null Checks are missing at various places, can you please add it?

mtyin committed:
Address various review feedbacks

1) Fix style of python notebook;
2) GraphStore:
   - improve a corner case by avoiding full set iteration;
   - add some null checking;
   - move utilities around.

mtyin commented on docs/graph_store.ipynb:
```diff
@@ -0,0 +1,903 @@
+{
```
I made some changes but from what I can see, the other notebooks are also using brightwhite as background color. LMK if I missed anything.

mtyin committed:
Format the notebook

mtyin commented:
/gcbrun

averikitsch commented:
/gcbrun

mtyin committed:
Fix linter error

averikitsch commented:
/gcbrun

mtyin committed:
Fix type annotation

mtyin commented:
/gcbrun

mtyin commented:
/gcbrun

mtyin committed:
Use batch table insertion API instead of transaction DML API

averikitsch commented on requirements.txt:
```diff
@@ -1,3 +1,4 @@
 langchain-core==0.3.9
 langchain-community==0.3.1
 google-cloud-spanner==3.49.1
+langchain-experimental==0.3.2
```
Since this dep is only needed for the colab, please remove from here.

mtyin replied:
Done.

averikitsch commented on docs/graph_store.ipynb:
```diff
@@ -0,0 +1,923 @@
+    "colab_type": "text"
+   },
+   "source": [
+    "<a href=\"https://colab.research.google.com/gist/mtyin/3f1e7d8e4c6b59edc6e7858c52247465/copy-of-graph_store.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```
please remove duplicate colab button

mtyin replied:
Good catch! Done.

mtyin committed:
Minor fixes & cleanups: colab, requirements, tests, schema corner cases

mtyin commented:
> Thank you [mtyin](https://github.com/mtyin) for raising the PR. The code looks well organized and modular.
> 
> I have reviewed the source code file and not test file. From the functionality point of view, it looks good but I need to do one more iteration.
> 
> Other than mentioned comments, few general things:
> 
> 1. You can use `isort` and `black` to format and style the code.
> 2. Null Checks are missing at various places, can you please add it?

Resolve this.

gauravpurohit06 reviewed (approved):
LGTM
